pkg/client/apiutil/lazyrestmapper.go

```diff
@@ -0,0 +1,260 @@
+*/
+
+package apiutil
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// LazyRESTMapper is a RESTMapper that will lazily query the provided
+// client for discovery information to do REST mappings.
+type LazyRESTMapper struct {
+	mapper      meta.RESTMapper
+	client      *discovery.DiscoveryClient
+	knownGroups []string
+}
+
+// NewLazyRESTMapper initializes a LazyRESTMapper
+func NewLazyRESTMapper(c *rest.Config) (meta.RESTMapper, error) {
+	dc, err := discovery.NewDiscoveryClientForConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return &LazyRESTMapper{
+		mapper: restmapper.NewDiscoveryRESTMapper([]*restmapper.APIGroupResources{}),
+		client: dc,
+	}, nil
+}
+
+func (m *LazyRESTMapper) addKnownGroupAndReload(groupName string) error {
+	m.knownGroups = append(m.knownGroups, groupName)
+
+	groupResources, err := m.getFilteredAPIGroupResources(m.client)
+	if err != nil {
+		return err
+	}
+
+	m.mapper = restmapper.NewDiscoveryRESTMapper(groupResources)
+
+	return nil
+}
+
+// KindFor implements Mapper.KindFor
+func (m *LazyRESTMapper) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
+	res, err := m.mapper.KindFor(resource)
+	if meta.IsNoMatchError(err) {
+		if err = m.addKnownGroupAndReload(resource.Group); err != nil {
+			return res, err
+		}
+		res, err = m.mapper.KindFor(resource)
+	}
+	return res, err
+}
+
+// KindsFor implements Mapper.KindsFor
+func (m *LazyRESTMapper) KindsFor(resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error) {
+	res, err := m.mapper.KindsFor(resource)
+	if meta.IsNoMatchError(err) {
+		if err = m.addKnownGroupAndReload(resource.Group); err != nil {
+			return res, err
+		}
+		res, err = m.mapper.KindsFor(resource)
+	}
+	return res, err
+}
+
+// ResourceFor implements Mapper.ResourceFor
+func (m *LazyRESTMapper) ResourceFor(input schema.GroupVersionResource) (schema.GroupVersionResource, error) {
+	res, err := m.mapper.ResourceFor(input)
+	if meta.IsNoMatchError(err) {
+		if err = m.addKnownGroupAndReload(input.Group); err != nil {
+			return res, err
+		}
+		res, err = m.mapper.ResourceFor(input)
+	}
+	return res, err
+}
+
+// ResourcesFor implements Mapper.ResourcesFor
+func (m *LazyRESTMapper) ResourcesFor(input schema.GroupVersionResource) ([]schema.GroupVersionResource, error) {
+	res, err := m.mapper.ResourcesFor(input)
+	if meta.IsNoMatchError(err) {
+		if err = m.addKnownGroupAndReload(input.Group); err != nil {
+			return res, err
+		}
+		res, err = m.mapper.ResourcesFor(input)
+	}
+	return res, err
+}
+
+// RESTMapping implements Mapper.RESTMapping
+func (m *LazyRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*meta.RESTMapping, error) {
+	res, err := m.mapper.RESTMapping(gk, versions...)
```

Comment:
This method on the discovery restmapper fetches all api groups, which means this change doesn't make anything more lazy than it is today. Before we merge anything in this area I expect an integration test that verifies that we do exactly one api call for discovery at the time we first use a given group. And I recommend you write that test first to avoid going down more avenues that look like they would solve the problem but don't.